### PR TITLE
New methods to monitor Argo workflows

### PR DESCRIPTION
**Pull request by pacospace**
Signed-off-by: Francesco Murdaca <fmurdaca@redhat.com>

Example:
```
namespace: thoth-prod-tensorflow
label_selector: component=adviser
workflows: {'Failed': 177, 'Error': 53}
pods: {'trigger-finished-webhook': {'Succeeded': 177, 'Error': 1}, 'graph-sync-advise': {'Failed': 177, 'Error': 1}, 'advise': {'Succeeded': 177, 'Error': 52, 'Failed': 1}}
```

### CODE REVIEW

File: thoth/common/workflows.py [hunk 1]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
         template = self.openshift.oc_process(namespace, template)
         return template
 
+    def get_workflow(
+        self, namespace: str, name: str,
+    ):
```
Comment (fridex) [inline thoth/common/workflows.py:193]:
```
    ) -> Dict[str, Any]:
```

Missing here and in the other method signatures as well.

File: thoth/common/workflows.py [hunk 2]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
...
+        """Get Workflow in namespace by name."""
+        response = self.api.get_namespaced_workflow(namespace=namespace, name=name)
+        return response.to_dict()
+
+    def get_workflows(
+        self,
+        namespace: str,
+        label_selector: str,
+    ) -> dict:
```
Comment (fridex) [inline thoth/common/workflows.py:202]:
```
    ) -> Dict[str, Any]:
```

File: thoth/common/workflows.py [hunk 3]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
...
+        self, namespace: str, name: str,
+    ):
+        """Get Workflow in namespace by name."""
+        response = self.api.get_namespaced_workflow(namespace=namespace, name=name)
+        return response.to_dict()
+
+    def get_workflows(
+        self,
+        namespace: str,
+        label_selector: str,
```
Comment (fridex) [inline thoth/common/workflows.py:201]: Can be the label selector optional?

```
        label_selector: Optional[str] = None,
```

File: thoth/common/workflows.py [hunk 4]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
...
+    def get_workflows_info(
+        self, label_selector: str, namespace: str
+    ) -> Dict[str, int]:
+        """Get workflows info from a specific namespace."""
+        workflows = self.get_workflows(namespace=namespace, label_selector=label_selector)
+        workflow_data = {}
+        for workflow in workflows["items"]:
+
+            workflow_main = self._collect_workflow_info(workflow=workflow)
+
```
Comment (fridex) [inline thoth/common/workflows.py:224]: New lines not needed. You can use black to format the code.

File: thoth/common/workflows.py [hunk 5]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
...
+        for template in templates:
+            # TODO: Extend to other structures beside dag and consider also steps
+            if 'dag' in template.keys():
+                for task in template['dag']['tasks']:
+                    tasks_names.append(task['name'])
+        return tasks_names
+
+    @staticmethod
+    def _collect_workflows_tasks_info(nodes: dict, tasks_names: List[str]) -> dict:
```
Comment (fridex) [inline thoth/common/workflows.py:266]:
```
    def _collect_workflows_tasks_info(nodes: Dict[str, Any], tasks_names: List[str]) -> Dict[str, Any]:
```

File: thoth/common/workflows.py [hunk 6]
```diff
@@ -188,6 +188,162 @@ def get_workflow_template(
...
+        nodes_info = {}
+        if not nodes:
+            return nodes_info
+
+        for pod_id, node in nodes.items():
+            
+            if node['display_name'] in tasks_names:
+                nodes_info[pod_id] = {}
+                nodes_info[pod_id]['pod_id'] = pod_id
+                nodes_info[pod_id]['display_name'] = node['display_name']
```
Comment (fridex) [inline thoth/common/workflows.py:277]: This entry can be initialized at the end with all the values to safe some CPU cycles:

```
                nodes_info[pod_id] = {
                  "pod_id": pod_id,
                  "display_name": node["display_name"],
                  "finished_at": finished_at,
                   ...
                }
```

File: thoth/common/workflows.py (hunk 3 above)
Comment (pacospace) [reply thoth/common/workflows.py:201]: Yes it can be :)

**Comment by pacospace** [comment]
> Signed-off-by: Francesco Murdaca [fmurdaca@redhat.com](mailto:fmurdaca@redhat.com)
> 
> Example:
> 
> ```
> namespace: thoth-prod-tensorflow
> label_selector: component=adviser
> workflows: {'Failed': 177, 'Error': 53}
> pods: {'trigger-finished-webhook': {'Succeeded': 177, 'Error': 1}, 'graph-sync-advise': {'Failed': 177, 'Error': 1}, 'advise': {'Succeeded': 177, 'Error': 52, 'Failed': 1}}
> ```

Updated:

```
namespace: thoth-prod-tensorflow
label_selector: None
status: {'adviser': {'component': 'adviser', 'status': {'Failed': 182, 'Error': 53}, 'tasks': {'trigger-finished-webhook': {'Succeeded': 182, 'Error': 1}, 'graph-sync-advise': {'Failed': 182, 'Error': 1}, 'advise': {'Succeeded': 182, 'Error': 52, 'Failed': 1}}}, 'qeb-hwt': {'component': 'qeb-hwt', 'status': {'Error': 39, 'Failed': 3}, 'tasks': {'thamos-advise': {'Error': 39, 'Failed': 3}}}}
```

**Pull request merged**